source/extensions/common/aws/credentials_provider_impl.cc

```diff
@@ -62,13 +72,93 @@ Credentials EnvironmentCredentialsProvider::getCredentials() {
   return Credentials(access_key_id, secret_access_key, session_token);
 }
 
-void MetadataCredentialsProviderBase::refreshIfNeeded() {
+void CachedCredentialsProviderBase::refreshIfNeeded() {
   const Thread::LockGuard lock(lock_);
   if (needsRefresh()) {
     refresh();
   }
 }
 
+bool CredentialsFileCredentialsProvider::needsRefresh() {
+  return api_.timeSource().systemTime() - last_updated_ > REFRESH_INTERVAL;
+}
+
+std::string getEnvironmentVariableOrDefault(const std::string& variable_name,
+                                            const std::string& default_value) {
+  const char* value = getenv(variable_name.c_str());
+  return (value != nullptr) && (value[0] != '\0') ? value : default_value;
+}
+
+void CredentialsFileCredentialsProvider::refresh() {
+  ENVOY_LOG(debug, "Getting AWS credentials from the credentials file");
+
+  const auto credentials_file = getEnvironmentVariableOrDefault(
+      AWS_SHARED_CREDENTIALS_FILE, DEFAULT_AWS_SHARED_CREDENTIALS_FILE);
+  const auto profile = getEnvironmentVariableOrDefault(AWS_PROFILE, DEFAULT_AWS_PROFILE);
+
+  extractCredentials(credentials_file, profile);
+}
+
+void CredentialsFileCredentialsProvider::extractCredentials(const std::string& credentials_file,
+                                                            const std::string& profile) {
+  // Update last_updated_ now so that even if this function returns before successfully
+  // extracting credentials, we'll cache an empty set of credentials. This prevents envoy
```

Comment:
technically we are not caching an empty set of credentials. We just return without setting anything on `cached_credentials_`.

Do we want to mention that after `REFRESH_INTERVAL` (1 hour) period an attempt will be made to fetch credentials again from the credentials file?

Reply:
This has been addressed.